metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/HiveConnectorFastTableService.java

```diff
@@ -114,43 +84,67 @@ public boolean exists(final ConnectorRequestContext requestContext, final Qualif
         final List<String> uris,
         final boolean prefixSearch
     ) {
-        final long start = registry.clock().wallTime();
-        // Create the sql
-        final StringBuilder queryBuilder = new StringBuilder(SQL_GET_TABLE_NAMES_BY_URI);
-        final List<SqlParameterValue> params = Lists.newArrayList();
-        if (prefixSearch) {
-            queryBuilder.append(" and (1=0");
-            uris.forEach(uri -> {
-                queryBuilder.append(" or location like ?");
-                params.add(new SqlParameterValue(Types.VARCHAR, uri + "%"));
-            });
-            queryBuilder.append(" )");
+        return directSqlTable.getTableNames(uris, prefixSearch);
+    }
+
+    /**
+     * Update a table with the given metadata.
+     *
+     * If table is an iceberg table, then lock the table for update so that no other request can update it. If the meta
+     * information is invalid, then throw an error.
+     * If table is not an iceberg table, then do a regular table update.
+     *
+     * @param requestContext The request context
+     * @param tableInfo      The resource metadata
+     */
+    @Override
+    public void update(final ConnectorRequestContext requestContext, final TableInfo tableInfo) {
+        if (isIcebergTable(tableInfo)) {
+            final QualifiedName tableName = tableInfo.getName();
+            final Long tableId = directSqlTable.getTableId(tableName);
+            try {
+                log.debug("Locking Iceberg table {}", tableName);
+                directSqlTable.lockIcebergTable(tableId, tableName);
+                try {
+                    final TableInfo existingTableInfo = get(requestContext, tableInfo.getName());
+                    if (!isValidIcebergUpdate(existingTableInfo, tableInfo)) {
+                        throw new IllegalStateException("Invalid iceberg table metadata.");
```

Comment:
We should mark this as invalidMetaDataException, and separate this from real IllegalStateException ( locking exception ).

Reply:
I think you mark it as IllegalStateException is because of it's part of the check-put condition check. This is different from the real metadata validation check.  We should put the explicit message in the exception - "the existing metadata location does not match".